File: packages/sdk-ts/test/idempotency.test.ts
import { test, describe, beforeEach } from 'node:test';
import assert from 'node:assert/strict';
import { WixzelPhone, lastResponse } from '../src/index.js';
import { apiError, fakeFetch } from './helpers/fake-fetch.js';

const api = fakeFetch();
const client = new WixzelPhone({ apiKey: 'wv_live_abc', baseUrl: 'https://api.example', fetch: api.fetchImpl, sleep: api.sleep });

beforeEach(() => api.reset());

describe('idempotency', () => {
    test('calls.create and billing.createTopup send a generated key; other creates do not', async () => {
        await client.calls.create({ to: '+14155551234', agent_id: 'ag1' });
        assert.match(api.last().headers['idempotency-key']!, /^[0-9a-f-]{36}$/);

        await client.billing.createTopup({ amount_usd: 20 });
        assert.match(api.last().headers['idempotency-key']!, /^[0-9a-f-]{36}$/);

        await client.agents.create({ name: 'a', system_prompt: 'b', opening_message: 'c', voice: { realtime: { model: 'google/gemini-live' } } });
        assert.equal(api.last().headers['idempotency-key'], undefined);
    });

    test('an explicit key is sent verbatim and stays out of the body', async () => {
        await client.calls.create({ to: '+14155551234', agent_id: 'ag1' }, { idempotencyKey: 'order-42' });
        const req = api.last();
        assert.equal(req.headers['idempotency-key'], 'order-42');
        assert.deepEqual(req.body, { to: '+14155551234', agent_id: 'ag1' });
    });

    test('the same key is reused across a 429 retry', async () => {
        api.queue.push({ ...apiError(429, 'rate_limit_error', 'rate_limit_exceeded'), headers: { 'retry-after': '1' } }, { status: 201, body: { id: 'c1' } });
        await client.calls.create({ to: '+14155551234', agent_id: 'ag1' });
        assert.equal(api.requests.length, 2);
        assert.equal(api.requests[0]!.headers['idempotency-key'], api.requests[1]!.headers['idempotency-key']);
    });

    test('a replayed response is visible through lastResponse', async () => {
        api.queue.push({ status: 201, body: { id: 'c1' }, headers: { 'idempotent-replay': 'true' } });
        const call = await client.calls.create({ to: '+14155551234', agent_id: 'ag1' }, { idempotencyKey: 'k' });
        assert.equal(lastResponse(call)?.idempotentReplay, true);
        api.queue.push({ status: 201, body: { id: 'c2' } });
        const fresh = await client.calls.create({ to: '+14155551234', agent_id: 'ag1' });
        assert.equal(lastResponse(fresh)?.idempotentReplay, false);
    });

    test('a key over 255 characters is refused before any request', async () => {
        await assert.rejects(
            client.calls.create({ to: '+14155551234', agent_id: 'ag1' }, { idempotencyKey: 'x'.repeat(256) }),
            /1–255 characters/,
        );
        assert.equal(api.requests.length, 0);
    });
});

File: packages/sdk-dart/test/idempotency_test.dart
import 'package:test/test.dart';
import 'package:wixzel_phone/wixzel_phone.dart';

import 'helpers.dart';

void main() {
  final api = FakeApi();
  setUp(api.reset);

  group('idempotency', () {
    test('the money paths send a generated key; other creates do not',
        () async {
      final client = api.wixzel();

      await client.calls
          .create(const CreateCall(to: '+14155551234', agentId: 'ag1'));
      expect(api.last.headers['idempotency-key'], matches(r'^[0-9a-f-]{36}$'));

      await client.billing.createTopup(const CreateTopup(amountUsd: 20));
      expect(api.last.headers['idempotency-key'], matches(r'^[0-9a-f-]{36}$'));

      await client.agents.create(CreateAgent(
        name: 'a',
        systemPrompt: 'b',
        openingMessage: 'c',
        voice: VoiceConfig.realtimeModel(
          const RealtimeConfig(model: 'google/gemini-live'),
        ),
      ));
      expect(api.last.headers.containsKey('idempotency-key'), isFalse);
    });

    test('an explicit key is sent verbatim and stays out of the body',
        () async {
      final client = api.wixzel();
      await client.calls.create(
        const CreateCall(to: '+14155551234', agentId: 'ag1'),
        idempotencyKey: 'order-42',
      );
      expect(api.last.headers['idempotency-key'], 'order-42');
      expect(api.last.body, {'to': '+14155551234', 'agent_id': 'ag1'});
    });

    test('a replay is visible through createWithResponse', () async {
      final client = api.wixzel();
      api.queue.add(Scripted(
        status: 201,
        body: {'id': 'c1', 'session_id': 's1'},
        headers: {'idempotent-replay': 'true', 'x-request-id': 'req_9'},
      ));
      final replayed = await client.calls.createWithResponse(
        const CreateCall(to: '+14155551234', agentId: 'ag1'),
        idempotencyKey: 'k',
      );
      expect(replayed.idempotentReplay, isTrue);
      expect(replayed.requestId, 'req_9');
      expect(replayed.data.id, 'c1');

      api.queue
          .add(Scripted(status: 201, body: {'id': 'c2', 'session_id': 's2'}));
      final fresh = await client.calls.createWithResponse(
        const CreateCall(to: '+14155551234', agentId: 'ag1'),
      );
      expect(fresh.idempotentReplay, isFalse);
    });

    test('a key over 255 characters is refused before any request', () async {
      final client = api.wixzel();
      expect(
        () => client.calls.create(
          const CreateCall(to: '+14155551234', agentId: 'ag1'),
          idempotencyKey: 'x' * 256,
        ),
        throwsA(isA<ArgumentError>()),
      );
      expect(api.requests, isEmpty);
    });
  });
}
